Subject: [PATCH] network_backends.h: support darwin sendfile

The FreeBSD version of sendfile is already supported.  Starting
with OS X 10.5, Darwin also supports sendfile, but using a
slightly different argument list.

Add support for darwin's sendfile by introducing a suitable
macro to adjust the arguments.

Signed-off-by: Kyle J. McKay <mackyle@gmail.com>

---
 src/network_backends.h | 16 +++++++++++++++-
 1 file changed, 15 insertions(+), 1 deletion(-)

diff --git a/src/network_backends.h b/src/network_backends.h
index 8233c488..3b98c3d9 100644
--- a/src/network_backends.h
+++ b/src/network_backends.h
@@ -17,12 +17,26 @@
 # define USE_LINUX_SENDFILE
 #endif
 
-#if defined HAVE_SENDFILE && (defined(__FreeBSD__) || defined(__DragonFly__))
+#if defined HAVE_SENDFILE && (defined(__FreeBSD__) || defined(__DragonFly__) || defined(__APPLE__))
 # ifdef USE_SENDFILE
 #  error "can't have more than one sendfile implementation"
 # endif
 # define USE_SENDFILE "freebsd-sendfile"
 # define USE_FREEBSD_SENDFILE
+# include <sys/uio.h>
+# ifdef __APPLE__
+#  include <sys/types.h>
+#  include <sys/socket.h>
+#  include <limits.h>
+#  ifndef UIO_MAXIOV
+#   define UIO_MAXIOV IOV_MAX
+#  endif
+   static __inline int _darwin_sendfile(
+    int fd, int s, off_t offset, off_t *len, struct sf_hdtr *hdtr, int flags)
+    {return sendfile(fd, s, offset, len, hdtr, flags);}
+#  define sendfile(fd, s, offset, nbytes, hdtr, sbytes, flags) \
+    ((*sbytes)=(off_t)(nbytes),_darwin_sendfile(fd, s, offset, sbytes, hdtr, flags))
+# endif /* __APPLE__ */
 #endif
 
 #if defined HAVE_SYS_SENDFILE_H && defined HAVE_SENDFILEV && defined(__sun)
---
